cpp/src/arrow/array.cc

```diff
@@ -375,6 +375,14 @@ std::shared_ptr<Array> UnionArray::child(int i) const {
   return boxed_fields_[i];
 }
 
+const Array* UnionArray::UnsafeChild(int i) const {
+  if (!boxed_fields_[i]) {
+    DCHECK(MakeArray(data_->child_data[i], &boxed_fields_[i]).ok());
```

Comment:
I am not totally sure this code gets executed in release builds, taking a look

Reply:
I think the compiler is able to reason about the fact that this has side effects.

Reply:
Hm, interesting, I just copied the code from UnionArray::child. Do you think it would be better to do it in two lines and first call MakeArray and then DCHECK for the return Status to be OK?

Reply:
I added this now, feel free to revert if you prefer the other way.